### PR TITLE
[AdminBundle] Mark classes internal/final

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves partially #7189

### CODE REVIEW

File: bundles/AdminBundle/Controller/Admin/Document/PageController.php [hunk 1]
```diff
@@ -26,6 +26,8 @@
 
 /**
  * @Route("/page")
+ *
+ * @internal
  */
 class PageController extends DocumentControllerBase
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Admin/Document/PageController.php:32]:
```suggestion
final class PageController extends DocumentControllerBase
```

File: bundles/AdminBundle/Controller/Admin/Document/PrintpageControllerBase.php [hunk 2]
```diff
@@ -22,6 +22,9 @@
 use Symfony\Component\HttpFoundation\JsonResponse;
 use Symfony\Component\HttpFoundation\Request;
 
+/**
+ * @internal
+ */
 class PrintpageControllerBase extends DocumentControllerBase
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Admin/Document/PrintpageControllerBase.php:28]:
```suggestion
abstract class PrintpageControllerBase extends DocumentControllerBase
```

File: bundles/AdminBundle/Controller/Admin/ElementControllerBase.php [hunk 3]
```diff
@@ -32,6 +32,9 @@
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
 
+/**
+ * @internal
+ */
 class ElementControllerBase extends AdminController
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Admin/ElementControllerBase.php:38]:
```suggestion
abstract class ElementControllerBase extends AdminController
```

File: bundles/AdminBundle/Controller/GDPR/PimcoreUsersController.php [hunk 4]
```diff
@@ -27,8 +27,10 @@
  * @Route("/pimcore-users")
  *
  * @package GDPRDataExtractorBundle\Controller
```
Comment (brusch) [inline bundles/AdminBundle/Controller/GDPR/PimcoreUsersController.php:29]:
```suggestion
```

File: bundles/AdminBundle/Controller/GDPR/SentMailController.php [hunk 5]
```diff
@@ -27,8 +27,10 @@
  * @Route("/sent-mail")
  *
  * @package GDPRDataExtractorBundle\Controller
```
Comment (brusch) [inline bundles/AdminBundle/Controller/GDPR/SentMailController.php:29]:
```suggestion
```

File: bundles/AdminBundle/Controller/Reports/AnalyticsController.php [hunk 6]
```diff
@@ -27,6 +27,8 @@
 
 /**
  * @Route("/analytics")
+ *
+ * @internal
  */
 class AnalyticsController extends ReportsControllerBase implements KernelControllerEventInterface
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Reports/AnalyticsController.php:33]:
```suggestion
final class AnalyticsController extends ReportsControllerBase implements KernelControllerEventInterface
```

File: bundles/AdminBundle/Controller/Reports/CustomReportController.php [hunk 7]
```diff
@@ -24,6 +24,8 @@
 
 /**
  * @Route("/custom-report")
+ *
+ * @internal
  */
 class CustomReportController extends ReportsControllerBase
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Reports/CustomReportController.php:30]:
```suggestion
final class CustomReportController extends ReportsControllerBase
```

File: bundles/AdminBundle/Controller/Reports/ReportsControllerBase.php [hunk 8]
```diff
@@ -17,6 +17,9 @@
 use Pimcore\Bundle\AdminBundle\Controller\AdminController;
 use Pimcore\Config;
 
+/**
+ * @internal
+ */
 class ReportsControllerBase extends AdminController
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Reports/ReportsControllerBase.php:23]:
```suggestion
abstract class ReportsControllerBase extends AdminController
```

File: bundles/AdminBundle/Controller/Reports/SettingsController.php [hunk 9]
```diff
@@ -22,6 +22,8 @@
 
 /**
  * @Route("/settings")
+ *
+ * @internal
  */
 class SettingsController extends ReportsControllerBase
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Reports/SettingsController.php:28]:
```suggestion
final class SettingsController extends ReportsControllerBase
```

File: bundles/AdminBundle/EventListener/CustomAdminEntryPointCheckListener.php [hunk 10]
```diff
@@ -24,6 +24,9 @@
 use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
 use Symfony\Component\HttpKernel\KernelEvents;
 
+/**
+ * @internal
+ */
 class CustomAdminEntryPointCheckListener implements EventSubscriberInterface
```
Comment (brusch) [inline bundles/AdminBundle/EventListener/CustomAdminEntryPointCheckListener.php:30]:
```suggestion
final class CustomAdminEntryPointCheckListener implements EventSubscriberInterface
```

**Review by brusch — CHANGES_REQUESTED** [review]

**Review by brusch — APPROVED** [review]

**Comment by roland4432** [comment]
What's the reason for all classes to be marked as final?

**Comment by brusch** [comment]
@roland4432 Because they are not intended to be overridden or customized. The BC promise / SemVer doesn't apply to those classes.

**Comment by roland4432** [comment]
Could you maybe leave that decission to the programmer?
This breaks dozens of projects and doesn't make any sense in the symfony flow at all.

And.. it doesn't even have the desired effect, since you can still override the controller just not extend it.

**Comment by roland4432** [comment]
I am interested what top contributors like @BlackbitNeueMedien @dpfaffenbauer @jdreesen think about that...

**Comment by BlackbitNeueMedien** [comment]
I never understood the argument to make classes final. Arguments like 
> Using the class for something else by overriding part of its behavior may jeopardize its integrity. The state of the extended object may become unpredictable or inconsistent with the state of the original object.
from https://matthiasnoback.nl/2018/09/final-classes-by-default-why/

are simply not true. If you want to protect internal state of properties, then it is fine for me to make them private. But to prohibit extending a class will only lead to a lot of copy-paste. There are a lot of use-cases where reusing / extending the AdminBundle controllers will on the one hand make custom features easier to implement and on the other hand make them easier to migrate. Because what will happen is that all the custom features which currently extend the AdminBundle controllers which @roland4432 mentions will have to copy the methods from the original classes to still work. Then this at first will work but when those AdminBundle classes get changed, the copied methods might again not work. Of course now you will say that developers have to check their code anyway for Pimcore updates - but what is the advantage of making those classes final?

**Comment by brusch** [comment]
For me it would be ok as well to just tag them as `@internal`, but that would still mean that we're not maintaining backward-compatibility for those classes - "use at your own risk".

**Comment by roland4432** [comment]
Would be totally fine with that.

**Pull request merged**